app/api/services/users/patch_user.py

```diff
@@ -1,25 +1,22 @@
-import dataclasses
 from datetime import date
-from typing import Optional
+from typing import TypedDict
 
 import apiflask
 from sqlalchemy import orm
 
 from api.db.models.user_models import Role, User
 from api.route.api_context import ApiContext
-from api.services.core.patch_params import Missing, fields_to_patch, missing
 from api.services.users.create_user import RoleParams
 
 
-@dataclasses.dataclass
-class PatchUserParams:
-    first_name: str | Missing = missing
-    middle_name: str | Missing = missing
-    last_name: str | Missing = missing
-    phone_number: str | Missing = missing
-    date_of_birth: date | Missing = missing
-    is_active: bool | Missing = missing
-    roles: list[RoleParams] | Missing = missing
+class PatchUserParams(TypedDict, total=False):
+    first_name: str
+    middle_name: str
+    last_name: str
+    phone_number: str
+    date_of_birth: date
+    is_active: bool
+    roles: list[RoleParams]
```

Comment:
Did we not want to do the define-once approach with them differing on `total=False`? Ah, wait, the PEP explicitly states it has to be done that way as it's only there for backwards compatibility.

I did try tinkering with inheritance slightly. You can't do a base class without any inheritance and then try to do `class Subclass(BaseClass, TypedDict)`, multiple inheritance is explicitly blocked for TypedDict.

You also can't do:
```py
class BaseExample(TypedDict):
    x: str
    y: Optional[str]

class ExampleA(BaseExample, total=False):
    pass

base: BaseExample = {"x": "asdasdasd"}
something: ExampleA = {"x": "asdasd"}
```
as the `total=False` doesn't get set - and I don't know enough about the class-inheritance bits in the TypedDict typing meta functions it calls to know if there is a clever workaround (probably not, and probably bad to consider).

So, yeah, it'd need to be duplicated, but I think that's just how it needs to be regardless of the technical approach (unless you want to do a lot of not-null checks for POST).

Reply:
appreciate the stream of consciousness!

i like this approach best too, i'll go ahead and merge it into the other PR